### PR TITLE
bluetooth: controller: Implemented dynamic re-scheduling in ticker

### LINKED ISSUE #20264: Bluetooth: Delay advertising events instead of dropping them on collision
**Is your enhancement proposal related to a problem? Please describe.**
Currently, an advertising event is dropped when it collides in time with a higher priority event, unless a sufficient number of consecutive advertising events have already been dropped, in which case the advertising event wins, and the higher priority event is dropped. Such collisions can be quite frequent when advertising is running in parallel with multiple connections, especially when some of those connections are carrying audio in large packets. When many advertising events are dropped, it may impair the ability of other devices to detect and connect to the advertiser, and the user may experience this as undesirable latencies and sluggishness.

**Describe the solution you'd like**
Instead of dropping the advertising event when there's a collision, the event should be rescheduled to the first free time slot in the future. To keep things simple, a time slot should only be considered free if it is large enough to hold the advertising event without overlapping any other events, i.e. priorities/aging should not be considered even if this would allow the advertising event to win over another event in an earlier time slot.

The total time between advertising events should probably be restricted to T_advEvent = advInterval + advDelay as specified in the Core spec v5.1 vol 6 part B section 4.4.2.2.1, where advDelay is a random value in the range 0 - 10 ms. So, if the first free slot corresponds to an advDelay larger than 10 ms, the event should probably be dropped. In order to stay within this limit when rescheduling (which may happen multiple times for the same event), the scheduler would need to store the actual time limit for the given event, until the event is either successfully executed or dropped.

This solution should significantly reduce the number of dropped advertising events as well as the number of other events dropped due to the aging of advertising events, thus increasing the responsiveness of the overall system. The randomization of advDelay would no longer have a flat distribution - large values would have a slightly higher probability than small values - but this is probably a small price to pay.

**Describe alternatives you've considered**
As mentioned, the spec sets an upper limit of 10 ms on advDelay, so the proposal is to drop the event rather than exceed this limit. However, dropping the event is also in a way equivalent to exceeding the limit (by a much larger amount), so would it actually be better to just ignore the limit? A downside of doing so, is that there would be no theoretical upper bound on how far into the future to search for a free time slot.

Instead of always delaying the advertising event to avoid a collision, one could consider allowing it to be advanced in time as well. This would reduce the risk of dropping events that are already close to the advDelay limit. However, I'm told that collision detection currently happens quite late, so it is not possible to advance the time. Also, if advancing were to occur frequently, it might cause a noticeable increase in average power, which would violate the spec.

**Additional context**
None.

### PR DESCRIPTION
**Pull request by mtpr-ot**
As a part of a ticker extension interface, it is now possible to specify
a slot_window when starting a ticker. When setting the
ticks_slot_window to a non-zero value, it is requested that the node
timeout is re-located to a position within the window, where the node
does not collide with other nodes - aligning to the end of the window.
The solution takes into consideration if a node has already been
updated with drift correction (e.g. ADV randomization), subtracting this
from the window.

Fixes #20264

Signed-off-by: Morten Priess <mtpr@oticon.com>

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ull_adv.c [hunk 1]
```diff
@@ -879,7 +886,10 @@ u8_t ll_adv_enable(u8_t enable)
 #else
 				   TICKER_NULL_LAZY,
 #endif
-				   (adv->evt.ticks_slot + ticks_slot_overhead),
+				   ticks_slot,
+#if defined(CONFIG_BT_TICKER_EXT)
+				   HAL_TICKER_US_TO_TICKS(10000) + ticks_slot,
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ll_sw/ull_adv.c]: Should we have a define for 10000 ?

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 2]
```diff
@@ -82,18 +93,36 @@ struct ticker_node {
...
 /* User operation data structure for start opcode. Used for passing start
  * requests to ticker_job
  */
 struct ticker_user_op_start {
-	u32_t ticks_at_start;		/* Anchor ticks (absolute) */
-	u32_t ticks_first;		/* Initial timeout ticks */
-	u32_t ticks_periodic;		/* Ticker period ticks */
-	u32_t remainder_periodic;	/* Sub-microsecond tick remainder */
-	u16_t lazy;			/* Periodic latency in number of
+	u32_t  ticks_at_start;		/* Anchor ticks (absolute) */
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ticker/ticker.c]: Why the extra space added?

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 3]
```diff
@@ -62,6 +62,17 @@ struct ticker_node {
 					  * remainder
 					  */
 #if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE)
+#if  defined(CONFIG_BT_TICKER_EXT)
+	u32_t ticks_slot_window;	 /* Window in which the slot
```
Comment (cvinayak) [inline subsys/bluetooth/controller/ticker/ticker.c]: Can we get these 9 bytes (12 bytes unpacked) of memory supplied by user of the extension feature? Just have a pointer here to reference the memory defined by user/caller of ticker_start_ext interface.

This way only the advertiser will use 16 bytes extra, where as other states/roles use 4 bytes extra.

**Review by cvinayak — CHANGES_REQUESTED** [review]

**Comment by cvinayak** [comment]
samples/bluetooth/scan_adv, modified to use scan interval of 100ms and scan_window of 50 ms.

Before this PR, advertiser and scanner that overlapped skipped one of the state's event.

Before_commented

In comparison to ticker compatibility mode wherein collision mitigation was based on first-scheduled instance getting higher precedence, in the current ticker implementation current active/expired ticker instance gets precedence (which is not fair), and colliding instance needs to apply priority in order to grab the time space it is colliding with.

After this PR, advertiser pre-empts scanner on collision (without exercising priority based) and does not skip its time space. But on being done with its event, scanner is resumed with fresh scan window, this is not correct. 

After_commented

Is the new behavior as expected with this PR?

Thinking aloud, this PR should have slid the advertiser in +/- 10 ms window trying to be not skipped. But if it cant, it will have to skip without disturbing the scan window.

File: subsys/bluetooth/controller/ticker/ticker.c (hunk 2 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ticker/ticker.c]: > Why the extra space added?

Glitch ;-)

File: subsys/bluetooth/controller/ticker/ticker.c (hunk 3 above)
Comment (mtpr-ot) [reply subsys/bluetooth/controller/ticker/ticker.c]: Good idea

**Comment by mtpr-ot** [comment]
> In comparison to ticker compatibility mode wherein collision mitigation was based on first-scheduled instance getting higher precedence, in the current ticker implementation current active/expired ticker instance gets precedence (which is not fair), and colliding instance needs to apply priority in order to grab the time space it is colliding with.

According to the current ticker design (not compatibility mode), the expired ticker does not automatically get precedence. The "age" (highest period ticks) weighs equal to latency (lazy), and priority is only used if applied by host. This seems fair. If that's not what you see, I should investigate.

**Comment by mtpr-ot** [comment]
> After this PR, advertiser pre-empts scanner on collision (without exercising priority based) and does not skip its time space. But on being done with its event, scanner is resumed with fresh scan window, this is not correct.

No, this is not what I expect. If advertiser cannot find a slot within the 10 ms. window, it "dumps" the node at the end of its window, expecting a collision and being skipped. This is necessary for honoring "must expire". 
It is also supposed to have the advantage that if the node it collided with is updated, there is a chance it wil get air time anyway. If the ADV preempts the scanner, there's seems to be a bug. It should not be able to prempt anything.

**Comment by cvinayak** [comment]
> If that's not what you see, I should investigate.

This is not related to this PR, but something that I have noticed when using nRF flash driver that uses ticker at 7.5ms and a master connection establishment at 7.5 ms, that they dont round-robin and connection fails.

> If the ADV preempts the scanner, there seems to be a bug. It should not be able to preempt anything.

Yes, this seems to be a bug in this PR. One, ticker should not expire for the advertiser that overlaps the scanner; two, seem a secondary bug in LLL scan that it resumed with restored window (this is not observed otherwise when ticker does not expire an instance that overlaps scan window).

Hope you can address the issue "One".

**Comment by mtpr-ot** [comment]
Found the bug with preemption - the "previous" data was reset incorrectly, thereby forgetting that a node already reserved the slot. @cvinayak, please retest if you can find the time. Thanks.

**Comment by mtpr-ot** [comment]
> This is not related to this PR, but something that I have noticed when using nRF flash driver that uses ticker at 7.5ms and a master connection establishment at 7.5 ms, that they dont round-robin and connection fails.

I would like to investigate this. I have run several similar scenarios without problems, but maybe the nRF LLL makes the behavior different from what we see. Any idea how to reproduce in simulation?

**Comment by cvinayak** [comment]
> Found the bug with preemption - the "previous" data was reset prematurely, thereby forgetting that a node already reserved the slot.

I have tested this and the PR now does not introduce any regression in scan_adv sample.

**Comment by mtpr-ot** [comment]
> I have tested this and the PR now does not introduce any regression in scan_adv sample.

Thanks Vinayak.

**Comment by mtpr-ot** [comment]
Just did a DMEM profiling with and without BT_TICKER_EXT feature enabled.
My configuration includes CONFIG_BT_BROADCASTER, CONFIG_BT_OBSERVER and CONFIG_BT_MAX_CONN=5.

The added DMEM usage for BT_TICKER_EXT reported by our target compiler is **92 bytes**.

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 4]
```diff
@@ -791,17 +811,44 @@ void ticker_worker(void *param)
...
+			   (ticker->lazy_periodic <= ticker->lazy_current)) {
+				/* Mark node for re-scheduling in ticker_job */
+				ticker->ext_data->reschedule_state =
+					TICKER_RESCHEDULE_STATE_PENDING;
+			} else if (ticker->ext_data) {
+				/* Mark node as not re-scheduling */
+				ticker->ext_data->reschedule_state =
+					TICKER_RESCHEDULE_STATE_NONE;
+			}
+#endif /* CONFIG_BT_TICKER_EXT */
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Here a collision has been detected. It is checked if node has slot_window set, and if so set the reschedule state pending for ticker_job.

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 5]
```diff
@@ -791,17 +811,44 @@ void ticker_worker(void *param)
...
 			ticker->lazy_current++;
+
 			if ((ticker->must_expire == 0U) ||
-			    (ticker->lazy_periodic >= ticker->lazy_current)) {
-				/* Not a must-expire case or this is programmed
-				 * latency. Skip this ticker node
+			    (ticker->lazy_periodic >= ticker->lazy_current) ||
+			    TICKER_RESCHEDULE_PENDING(ticker)) {
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: We always skip timeout callback if rescheduling has become pending. Must-expire is handled at rescheduled time- also if there's no free slot found.

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 6]
```diff
@@ -1056,6 +1103,14 @@ static inline void ticker_job_node_update(struct ticker_node *ticker,
 	ticker->ticks_to_expire_minus +=
 				user_op->params.update.ticks_drift_minus;
 
+#if defined(CONFIG_BT_TICKER_EXT)
+	if (ticker->ext_data) {
+		ticker->ext_data->ticks_drift =
+			user_op->params.update.ticks_drift_plus -
+			user_op->params.update.ticks_drift_minus;
+	}
+#endif /* CONFIG_BT_TICKER_EXT */
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Store the applied drift - we must reduce the rescheduling window by this value.

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 7]
```diff
@@ -1303,11 +1359,17 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
...
+			 * the slot_previous information
+			 */
+			if (!TICKER_RESCHEDULE_PENDING(ticker)) {
+				/* save current ticks_slot_previous */
+				if (ticker->ticks_slot != 0U) {
+					instance->ticker_id_slot_previous =
+						id_expired;
+					instance->ticks_slot_previous =
+						ticker->ticks_slot;
+				}
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: This change to make sure that when rescheduling a node, the id_slot_previous and slot_previous information is not cleared, as we need it for the next ticker_worker.

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 8]
```diff
@@ -1317,11 +1379,21 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
...
+		    TICKER_RESCHEDULE_PENDING(ticker)) {
 #if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE)
-			/* Reload ticks_to_expire with one period */
-			ticker->ticks_to_expire  = ticker->ticks_periodic;
+			if (TICKER_RESCHEDULE_PENDING(ticker)) {
+				/* Set the re-scheduled node to now. Will be
+				 * collision resolved after all nodes are
+				 * restarted
+				 */
+				ticker->ticks_to_expire = ticks_elapsed;
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Just puts the reschedule-pending node in front for rescheduling in ticker_job.

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 9]
```diff
@@ -791,17 +811,44 @@ void ticker_worker(void *param)
 		 */
 		if (ticker->ticks_slot != 0U &&
 		   (slot_reserved || ticker_resolve_collision(node, ticker))) {
+#if defined(CONFIG_BT_TICKER_EXT)
+			if (ticker->ext_data &&
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Action: Dereference ticker->ext_data

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 10]
```diff
@@ -1303,11 +1359,17 @@ static inline void ticker_job_worker_bh(struct ticker_instance *instance,
...
 
-			/* save current ticks_slot_previous */
-			if (ticker->ticks_slot != 0U) {
-				instance->ticker_id_slot_previous = id_expired;
-				instance->ticks_slot_previous =
-					ticker->ticks_slot;
+			/* If a reschedule is set pending, we will need to keep
+			 * the slot_previous information
+			 */
+			if (!TICKER_RESCHEDULE_PENDING(ticker)) {
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Action: comine into one statement (AND)

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 11]
```diff
@@ -1432,7 +1507,214 @@ static inline u32_t ticker_job_insert(struct ticker_instance *instance,
...
+		}
+
+		/* Remove latency added in ticker_worker */
+		ticker->lazy_current--;
+
+		/* Prevent repeated re-scheduling */
+		ext_data->reschedule_state =
+			TICKER_RESCHEDULE_STATE_DONE;
+
+		rescheduling = 1U;
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Action: Remove

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 12]
```diff
@@ -1432,7 +1507,214 @@ static inline u32_t ticker_job_insert(struct ticker_instance *instance,
...
+		}
+
+		ticker_id_iter = nodes[ticker_id_head].next;
+		ticker_id_prev = TICKER_NULL;
+
+		/* If drift was applied to this node, this must be
+		 * taken into consideration. Reduce the window with
+		 * the amount of drift already applied.
+		 */
+		if (ext_data->ticks_drift < ext_data->ticks_slot_window) {
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Action: Add TODO regarding appending the random delay to the window...

File: subsys/bluetooth/controller/ticker/ticker.c [hunk 13]
```diff
@@ -1432,7 +1507,214 @@ static inline u32_t ticker_job_insert(struct ticker_instance *instance,
...
+				break;
+			}
+			ticker_id_prev = ticker_id_iter;
+			ticker_id_iter = node->next;
+		}
+
+		if (ticker_id_prev != TICKER_NULL) {
+			/* Node did not become the first - insert it */
+			instance->ticker_id_head   = nodes[ticker_id_head].next;
+			nodes[ticker_id_head].next = nodes[ticker_id_prev].next;
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/ticker/ticker.c]: Action: Add blank line to separate head update with re-linking. Second part comment: "Re-insert/link ticker nodes"

**Comment by mtpr-ot** [comment]
@cvinayak, PR now updated according to actions, please evaluate. Thanks.

**Comment by cvinayak** [comment]
scan_adv sample observations (mesh usecase):

Ticker compatibility,
image

Continuous scanning,
image

Continuous scanning with PR,
image

100ms interval and 50 ms window,
image

100ms interval and 50 ms window with PR,
image

Changes in this PR look good.

**Review by cvinayak — APPROVED** [review]

**Pull request merged**